### PR TITLE
Deeply deserialize groups

### LINKED ISSUE #6341: Chords contained in a group raise `AttributeError`s during freezing if passed through a backend
When calling `self.replace()` with a signature reconstructed from a serialized dictionary (e.g. after having been passed through a backend), if that signature is (or contains) a group which itself contains a chord, celery explodes after attempting to treat deeply nested dictionaries as signature objects. See below for a minimal repro I got together today.

My gut feel is that instantiating a signature from a dictionary may not be recursing down through the structure far enough and it leaves some of the encapsulated tasks as dicts. I've also noticed that groups containing a group also break in the same way, but I think that's because there's an internal promotion to a chord happening somewhere.

# Checklist

- [x] I have verified that the issue exists against the `master` branch of Celery.
- [ ] This has already been asked to the [discussion group](https://groups.google.com/forum/#!forum/celery-users) first.
- [x] I have read the relevant section in the
  [contribution guide](http://docs.celeryproject.org/en/latest/contributing.html#other-bugs)
  on reporting bugs.
- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Bug+Report%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical bug reports.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22PR+Type%3A+Bugfix%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed fixes.
- [x] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the bug was already fixed in the master branch.
- [x] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

## Mandatory Debugging Information

- [ ] I have included the output of ``celery -A proj report`` in the issue.
    (if you are not able to do this, then at least specify the Celery
     version affected).
- [ ] I have verified that the issue exists against the `master` branch of Celery.
- [ ] I have included the contents of ``pip freeze`` in the issue.
- [ ] I have included all the versions of all the external dependencies required
  to reproduce this bug.

## Optional Debugging Information

- [ ] I have tried reproducing the issue on more than one Python version
  and/or implementation.
- [ ] I have tried reproducing the issue on more than one message broker and/or
  result backend.
- [ ] I have tried reproducing the issue on more than one version of the message
  broker and/or result backend.
- [ ] I have tried reproducing the issue on more than one operating system.
- [ ] I have tried reproducing the issue on more than one workers pool.
- [ ] I have tried reproducing the issue with autoscaling, retries,
  ETA/Countdown & rate limits disabled.
- [ ] I have tried reproducing the issue after downgrading
  and/or upgrading Celery and its dependencies.

#### Related Issues
 - #4015

#### Possible Duplicates

- None

## Environment & Settings

**Celery version**:

<details>
<summary><b><code>celery report</code> Output:</b></summary>
<p>

```
```

</p>
</details>

# Steps to Reproduce

## Required Dependencies

* **Minimal Python Version**: N/A or Unknown
* **Minimal Celery Version**: N/A or Unknown
* **Minimal Kombu Version**: N/A or Unknown
* **Minimal Broker Version**: N/A or Unknown
* **Minimal Result Backend Version**: N/A or Unknown
* **Minimal OS and/or Kernel Version**: N/A or Unknown
* **Minimal Broker Client Version**: N/A or Unknown
* **Minimal Result Backend Client Version**: N/A or Unknown

### Python Packages

<details>
<summary><b><code>pip freeze</code> Output:</b></summary>
<p>

```
amqp==5.0.1
billiard==3.6.3.0
# Editable install with no version control (celery==5.0.0rc3)
-e /home/maybe/tmp/capp/venv/lib/python3.8/site-packages
click==7.1.2
click-didyoumean==0.0.3
click-repl==0.1.6
future==0.18.2
kombu==5.0.2
prompt-toolkit==3.0.7
pytz==2020.1
redis==3.5.3
six==1.15.0
vine==5.0.0
wcwidth==0.2.5
```

</p>
</details>

### Other Dependencies

<details>
<p>
N/A
</p>
</details>

## Minimally Reproducible Test Case

<details>
<p>

```python
import celery

app = celery.Celery("app", backend="redis://")

@app.task
def foo(*_):
    return 42

@app.task(bind=True)
def replace_with(self, sig):
    assert isinstance(sig, dict)
    sig = celery.Signature.from_dict(sig)
    raise self.replace(sig)

if __name__ == "__main__":
    sig = celery.group(
        celery.group(foo.s()),
    )
    res = sig.delay()
    print(res.get())

    sig.freeze()
    res = replace_with.delay(sig)
    print(res.get())
```

</p>
</details>

# Expected Behavior
It shouldn't explode. Presumably tasks within the group/chord should be signatures rather than dicts.

# Actual Behavior

Stack trace in the worker output:
```
[2020-09-08 12:44:05,453: DEBUG/MainProcess] Task accepted: app.replace_with[dcea02fd-23a3-404a-9fdd-b213eb51c0d1] pid:453431
[2020-09-08 12:44:05,457: ERROR/ForkPoolWorker-8] Task app.replace_with[dcea02fd-23a3-404a-9fdd-b213eb51c0d1] raised unexpected: AttributeError("'dict' object has no attribute '_app'")
Traceback (most recent call last):
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/kombu/utils/objects.py", line 41, in __get__
    return obj.__dict__[self.__name__]
KeyError: 'app'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/celery/app/trace.py", line 409, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/celery/app/trace.py", line 701, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/maybe/tmp/capp/app.py", line 13, in replace_with
    raise self.replace(sig)
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/celery/app/task.py", line 894, in replace
    sig.freeze(self.request.id)
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/celery/canvas.py", line 1302, in freeze
    self.tasks = group(self.tasks, app=self.app)
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/kombu/utils/objects.py", line 43, in __get__
    value = obj.__dict__[self.__name__] = self.__get(obj)
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/celery/canvas.py", line 1456, in app
    return self._get_app(self.body)
  File "/home/maybe/tmp/capp/venv/lib64/python3.8/site-packages/celery/canvas.py", line 1466, in _get_app
    app = tasks[0]._app
AttributeError: 'dict' object has no attribute '_app'
```

### PR DESCRIPTION
**Pull request by maybe-sybr**
## Description
Fixes #6341

### CODE REVIEW

**Comment by maybe-sybr** [comment]
Probably needs tests like the one that #4015 added but I need to switch off this for the moment

File: celery/canvas.py
Comment (maybe-sybr) [inline celery/canvas.py:1276]: My more complex failure case hits this code somewhere but it doesn't seem to be required to fix my breakage (just the change to groups). In any case, this seems like a legit change to make.

Comment (thedrow) [reply celery/canvas.py:1276]: Could you please include the more complicated cases as tests in our integration suite?
You can mark them as `xfail` for now.

**Review by thedrow — CHANGES_REQUESTED** [review]
This PR requires integration tests.

Please include the original test case in #6341 as a test in our integration suite.

File: celery/canvas.py
Comment (maybe-sybr) [reply celery/canvas.py]: Turns out this isn't strictly necessary because the `_maybe_group()` and `_maybe_signature()` in `chord.__init__()` cause nested `dicts` to get deserialized properly.

https://github.com/celery/celery/blob/9b5f6f5531f45525e904cef114d653a93f1c9635/celery/canvas.py#L1289-L1298

It's slightly different behaviour to the other `from_dict()` code paths since `chain.from_dict()` and (now) `group.from_dict()` deserialize their children prior to instantiating new objects. I've removed this from the diff, but let me know if you'd like it restored to make all three behave in the same way. (/ping @thedrow )

**Comment by maybe-sybr** [comment]
> This PR requires integration tests.
> 
> Please include the original test case in #6341 as a test in our integration suite.

@thedrow - sorry I took so long to get back to you on this. I've pushed up some unit tests which expose where the bug was actually manifesting (and importantly, where it wasn't - ie. in `chord`s). Per my other comment, I've removed the change to `chord.from_dict()` because the bug didn't manifest there , but it might still be nice to have that in the changeset so that all of the `from_dict()` implementations do the same thing, rather than relying on a side-effect of `chord.__init__()` which might change in future. Let me know what you think on that one.

I'm also happy to add an integration test based on the test case I posted in #6341 but I'm not sure how to go about that since I need to define some tasks to run with results being passed via some backend so that the serialization happens. The rough flow of the integration test would be:
1. for a task to create a group signature containing another group, and return that object.
1. another task to get the result of the first, and the run `Signature.from_dict()` on it
1. verification in the second task that the returned Signature object has indeed been deeply deseriaized

**Comment by maybe-sybr** [comment]
I noticed the test causing CI breakage [0] failing on my box too but thought it was just flakey. It's fine on master so definitely caused by my change - I'll sort it out now

[0] https://travis-ci.org/github/celery/celery/jobs/727590711

**Comment by maybe-sybr** [comment]
@thedrow - worked it out. Turns out the `group.freeze()` implementation relies on the sub-tasks it passes down for recursive freezing (via `group._freeze_unroll()`) being mutated in-place. By making a copy of the sub-tasks of a group encapsulated in a group (or a chord) when that sub-group made its way to `group.from_dict()` via `Signature.clone()`, it turns out I was creating a canonical version of the tasks in the sub-group, which the top level group wouldn't know about later. The frozen groupresult object looks like it was correct, but the top level group task wouldn't have been mutated like that test I broke assumed. Other bits of the codebase also seem to assume that things should be mutated in place if possible, so I've made an amendment to this diff to meet that assumption.

That's a bit complicated, so let me know if you want me to explain things any better. The tl;dr is that we didn't mutate a thing when we should have, and that broke assumptions. IMO the freezing logic could stand to be reworked a bit to get rid of some of these implicit mutations of built-in types like maps and lists, since they're a bit hard to follow - but that's certainly not the point here and I've just rolled with the current code expectations.

This should be good to merge now hopefully. I'll check in on CI as it runs though to make sure it's happy. Per my previous comment, still not sure about adding an integration test - let me know what you'd like to see in that space, if anything. Thanks!

**Comment by thedrow** [comment]
I've restarted the build to see if the error is consistent.

As for the integration tests, you should include your test case as a test in our suite.
If you have more complicated cases, please include them as well.

**Comment by maybe-sybr** [comment]
@thedrow - I think that CI failure is related to another odd behaviour I've observed with chains containing groups. I've spent some time tracing it and it looks like the change in 01dd66c (#5984 which fixed #5947 ) is causing this since it seems to cause an off-by-one in the calculated size of a chord body and the redis backend is never able to complete the chord.

The off by one reliably happens with my change to `group.from_dict()` whereas it doesn't happen on master for the same signature structure. Presumably because calling `maybe_signature()` if `group.from_dict()` is upgrading some nested signature structure into a chord and getting confused along the way. Commenting out the lines changed in #5984 makes the the broken test pass but obviously breaks the test added in that PR but obviously also breaks the test which was added. I'm continuing to try to work out what the right fix is here.

**Comment by maybe-sybr** [comment]
@thedrow - I think I've fixed the problems and have pushed up anther PR (#6354) which blocks this one (since this is based in that one). We can merge this one independently if we mark the failing test with an xfail which the other PR can resolve later. Let me know what you'd prefer.

**Review by auvipy — CHANGES_REQUESTED** [review]
the pytest is causing errors in integration tests. can you try to supress that?

**Comment by maybe-sybr** [comment]
per my comment on #6354 , I think those failures are on master and might be environmental rather than to do with actual code changes in celery :)

**Comment by maybe-sybr** [comment]
I've rebased this on top of the changeset for #6354 and hopefully it will also go green in CI. Please don't merge this until the other PR is merged :)

**Review by auvipy — CHANGES_REQUESTED** [review]
It will be easier to review after the prerequisite PR is merged

**Comment by maybe-sybr** [comment]
I've looked back over this and I think it may be possible to merge it independently. The changeset didn't really depend on anything in #6354 but I was under the impression that some of the integration testing would. I've rebased just the commits relevant to #6341 on top of master, so we'll see if CI is happy with it soon enough. In the meantime, you're welcome to review the diff again @thedrow and @auvipy

**Comment by maybe-sybr** [comment]
Ah, yes. That's the test I remember failing and sending me onto the chord counting tangent. I'll leave this as is for now so others can review the substantive diff, and rebase once #6354 gets merged.

**Review by auvipy — CHANGES_REQUESTED** [review]
please rebase

File: t/unit/tasks/test_canvas.py
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:850]: I think that this loop for example is a great opportunity to introduce [`pytest-subtests`](https://github.com/pytest-dev/pytest-subtests/) to our codebase.

There are other opportunities in this patch to use it.

File: t/unit/tasks/test_canvas.py
Comment (thedrow) [inline t/unit/tasks/test_canvas.py:899]:
```suggestion
        assert all(isinstance(task, Signature) for task in deserialized_chord.tasks)
```

That applies to all other loops of this type.

**Review by thedrow — CHANGES_REQUESTED** [review]

File: t/unit/tasks/test_canvas.py
Comment (maybe-sybr) [reply t/unit/tasks/test_canvas.py:850]: I've pivoted to using `all(...)` for the other loopy checks in this diff now, so this should be the only place subtests might be relevant. I'll do it here to support the first commit and introduce the subtest tooling so that it can be used elsewhere in the suite.

File: t/unit/tasks/test_canvas.py
Comment (maybe-sybr) [reply t/unit/tasks/test_canvas.py:899]: Done, this will land in my next push. I've used the trope throughout this test suite and updated two tests unrelated to this diff to match.

File: t/unit/tasks/test_canvas.py
Comment (maybe-sybr) [reply t/unit/tasks/test_canvas.py:850]: Oh, I actually see what you mean. We could use subtests to check, e.g., that the body of a deeply deserialized chord is correct, separate from the checks on the header! That's pretty cool, I'll do that too.

Comment (maybe-sybr) [reply t/unit/tasks/test_canvas.py:850]: Done, check out the new diff and let me know what you think

File: t/unit/tasks/test_canvas.py
Comment (maybe-sybr) [inline t/unit/tasks/test_canvas.py:1104]: I was unsure if this line should be part of the main test or a subtest. It's a prerequisite for the other subtests, so I feel like it might be better to take this out of a subtest block (and similarly in the tests below).

File: requirements/test-ci-base.txt
Comment (thedrow) [inline requirements/test-ci-base.txt:3]: Isn't this a duplicate requirement?
It's already specified in `test.txt`.

File: t/integration/tasks.py
Comment (thedrow) [inline t/integration/tasks.py:3]: Please run isort on this file.

File: t/unit/tasks/test_canvas.py
Comment (thedrow) [reply t/unit/tasks/test_canvas.py:1104]: Generally speaking if we follow a variation of the `AAA` pattern, every assert should be in a different subtest.

I'm not going to enforce that for the time being.

**Review by thedrow — CHANGES_REQUESTED** [review]
Just a few more minor things.

Overall this looks great!

File: t/unit/tasks/test_canvas.py
Comment (maybe-sybr) [reply t/unit/tasks/test_canvas.py:1104]: All good, sounds like I've done the right thing then for these tests with multiple assertions. `test_freeze_tasks_body_is_group()` has a single sanity assertion in the main body of the test but I'm okay with that structure since the rest of the test is nonsense if that is untrue.

File: requirements/test-ci-base.txt
Comment (maybe-sybr) [reply requirements/test-ci-base.txt:3]: Oh, I didn't realise that it only needed to be in one. I'll fix this.

File: t/integration/tasks.py
Comment (maybe-sybr) [reply t/integration/tasks.py:3]: `isort` seems fine with this file, however the import in `t/integration/test_canvas.py` needed to be moved. I'll amend and push in a moment.

**Comment by maybe-sybr** [comment]
Addressed those minor comments, sanity checked the `celery/` and `t/` trees with isort and it seems happy. Should be good to go, @thedrow - lmk if you see anything else!

**Review by auvipy — APPROVED** [review]
great job!

**Pull request merged**
